File: Sources/Redact/Redact.swift
import PlatformSupport

/// A redaction: text with unique placeholders, plus the mapping needed to
/// inspect the detections and restore the originals after out-of-band
/// processing (e.g. an LLM).
///
/// ```swift
/// let r = try await redact.redaction(of: userText)
/// let reply = try await llm.rewrite(r.redactedText)   // sees only [EMAIL_1], [GIVEN_NAME_1], …
/// let final = r.restore(reply)                          // originals filled back in
/// ```
public struct Redaction: Sendable {
    /// A single detected entity and the placeholder that stands in for it.
    public struct Item: Sendable, Identifiable, Equatable {
        public var id: String { placeholder }
        /// The category of personal data.
        public let label: Label
        /// The original (sensitive) text that was matched.
        public let original: String
        /// The unique, restorable placeholder, e.g. `"[EMAIL_1]"`.
        public let placeholder: String
        /// The model's confidence, `0...1` (deterministic recognizers report `1.0`).
        public let confidence: Double
        /// The entity's range in the *original* text.
        public let range: Range<String.Index>
    }

    /// The text with each entity replaced by its unique placeholder.
    public let redactedText: String
    /// Every detected entity, in document order.
    public let items: [Item]

    /// Fill the originals back into `processed` (typically an LLM's output) by
    /// substituting each placeholder. Placeholders are bracket-delimited and
    /// numbered (`[EMAIL_1]`, `[EMAIL_2]`, …) so they are unique and never a
    /// prefix of one another, so restoration is order-independent and safe.
    public func restore(_ processed: String) -> String {
        var out = processed
        for item in items { out = out.replacing(item.placeholder, with: item.original) }
        return out
    }
}

/// Options controlling detection and redaction.
public struct Options: Sendable {
    /// Minimum confidence for neural detections. Structured recognizers
    /// (email, cards, IBANs, …) always apply. Default `0.6`.
    public var minimumConfidence: Double
    /// If set, only these categories are detected/redacted; otherwise all are.
    public var labels: Set<Label>?

    public init(minimumConfidence: Double = 0.6, labels: Set<Label>? = nil) {
        self.minimumConfidence = minimumConfidence.isFinite
            ? min(1, max(0, minimumConfidence))
            : 0.6
        self.labels = labels
    }
}

/// Errors thrown while loading or running the model. (`MessageError` is
/// `LocalizedError` wherever Foundation exists, so `localizedDescription`
/// shows `message`.)
public enum RedactError: MessageError, Sendable {
    case resourceMissing
    case predictionFailed

    public var message: String {
        switch self {
        case .resourceMissing: "A Redact model resource was not found."
        case .predictionFailed: "On-device PII detection failed."
        }
    }
}

/// On-device, multilingual PII redaction.
///
/// `Redact` finds names, addresses, emails, phone numbers, cards, IBANs,
/// national IDs and more across the 24 official EU languages, fully on device.
/// Create one once and reuse it.
///
/// ```swift
/// let redact = Redact()
/// let r = try await redact.redaction(of: "Email Anna at anna@example.com.")
/// r.redactedText            // "Email [GIVEN_NAME_1] at [EMAIL_1]."
/// r.items.first?.original   // "Anna"
/// ```
public final class Redact: @unchecked Sendable {
    /// Resolve the model's assets (downloading/adopting as needed), reporting
    /// progress `0...1`.
    typealias ResolveAssets = @Sendable (@escaping @Sendable (Double) -> Void) async throws -> ModelAssets

    private let loader: LazyLoader<Model>
    private let availability: @Sendable () -> Bool

    /// Creates a redactor. Construction does no work and starts no download; the
    /// model loads on the first ``redaction(of:options:)`` or ``download(progress:)``,
    /// off your calling thread.
    ///
    /// `directory` is where the model lives. If it already contains the model
    /// (you pre-downloaded or shipped it there) it is used offline; otherwise
    /// the model is downloaded into it and reused offline afterward. With no
    /// `directory` (the default), a managed cache location is used.
    ///
    /// ```swift
    /// let redact = Redact()                       // download to the managed cache
    /// let redact = Redact(directory: myModelDir)  // the model lives here (use or download)
    /// let redact = Redact(bundle: myBundle)       // bundled in your app
    /// ```
    public convenience init(directory: String? = nil) {
        self.init(directory: directory, cacheRoot: nil)
    }

    /// Binding entry point that also supplies the platform base cache root under
    /// which the managed layout lives (the app cache dir on Android, node
    /// `~/.cache` on the web). On Apple/Linux FileManager provides it, so the
    /// public `init(directory:)` passes `nil`.
    @_spi(RedactBindings)
    public convenience init(directory: String?, cacheRoot: String?) {
        self.init(
            resolve: { try await Redact.resolvedAssets(directory: directory, cacheRoot: cacheRoot, progress: $0) },
            isAvailable: { Redact.isModelAvailable(directory: directory, cacheRoot: cacheRoot) }
        )
    }

    /// Creates a redactor from explicitly provided assets (used by the
    /// Android/JNI and custom-deployment paths).
    @_spi(RedactBindings)
    public convenience init(assets: ModelAssets) {
        self.init(resolve: { _ in assets }, isAvailable: { true })
    }

    init(resolve: @escaping ResolveAssets, isAvailable: @escaping @Sendable () -> Bool) {
        loader = LazyLoader { progress in try Model(assets: await resolve(progress)) }
        availability = isAvailable
    }

    /// Whether the model is available for this redactor with no network: cached
    /// (for the download source), present (for a directory), or bundled.
    public func isDownloaded() -> Bool { availability() }

    /// Download and load the model ahead of time, so the first
    /// ``redaction(of:options:)`` is instant. Reports download progress `0...1`.
    /// Concurrent calls, and an implicit load from a redaction, share one
    /// download. A no-op once loaded (see ``isDownloaded()``).
    public func download(progress: @Sendable @escaping (Double) -> Void = { _ in }) async throws {
        try await loader.run(progress: progress)
    }

    /// Await model readiness. The bindings use this to surface load errors
    /// eagerly; apps can just call ``redaction(of:options:)``.
    @_spi(RedactBindings)
    public func waitUntilLoaded() async throws {
        _ = try await loader.value()
    }

    /// Detect and redact the PII in `text`.
    ///
    /// The result's ``Redaction/redactedText`` has each entity replaced by a
    /// unique placeholder (`[EMAIL_1]`, `[GIVEN_NAME_1]`, …); ``Redaction/items``
    /// lists every detection (label, original text, placeholder, confidence,
    /// range); and ``Redaction/restore(_:)`` fills the originals back in.
    public func redaction(of text: String, options: Options = .init()) async throws -> Redaction {
        let spans = try await spans(in: text, options: options)
        let units = Array(text.utf16)
        var counts: [Label: Int] = [:]
        var items: [Redaction.Item] = []
        var out: [UInt16] = []
        var last = 0
        for span in spans {
            guard let label = Label(rawValue: span.label) else { continue }
            let s = span.start, e = span.end
            guard s >= last, s < e, e <= units.count else { continue }   // in order, non-overlapping
            let n = (counts[label] ?? 0) + 1
            counts[label] = n
            let placeholder = "[\(label.rawValue)_\(n)]"
            let original = String(decoding: units[s..<e], as: UTF16.self)
            let range = String.Index(utf16Offset: s, in: text)..<String.Index(utf16Offset: e, in: text)
            items.append(.init(label: label, original: original, placeholder: placeholder,
                               confidence: span.score, range: range))
            out.append(contentsOf: units[last..<s])
            out.append(contentsOf: Array(placeholder.utf16))
            last = e
        }
        out.append(contentsOf: units[last...])
        return Redaction(redactedText: String(decoding: out, as: UTF16.self), items: items)
    }

    private func spans(in text: String, options: Options) async throws -> [Span] {
        let model = try await loader.value()
        let allowed = options.labels.map { Set($0.map(\.rawValue)) }
        return try await model.detect(text, minScore: options.minimumConfidence)
            .filter { allowed?.contains($0.label) ?? true }
            .sorted { $0.start != $1.start ? $0.start < $1.start : $0.end < $1.end }
    }
}

File: packages/redact-kotlin/src/main/kotlin/ai/desertant/redact/Redact.kt
package ai.desertant.redact

import ai.desertant.core.FfiReader
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/** A single detected entity and the placeholder that stands in for it. */
data class RedactionItem(
    /** PII category, e.g. `"EMAIL"`. */
    val label: String,
    /** The original (sensitive) text that was matched. */
    val original: String,
    /** The unique, restorable placeholder, e.g. `"[EMAIL_1]"`. */
    val placeholder: String,
    /** Confidence in `0.0..1.0` (deterministic recognizers report `1.0`). */
    val confidence: Double,
    /** Start offset of [original] in the source text (UTF-16). */
    val start: Int,
    /** End offset (exclusive). */
    val end: Int,
)

/**
 * A redaction: text with unique placeholders, the detections, and the mapping
 * needed to restore the originals after out-of-band processing (e.g. an LLM).
 */
data class Redaction(
    /** The input with each entity replaced by its `[LABEL_N]` placeholder. */
    val redactedText: String,
    /** Every detected entity, in document order. */
    val items: List<RedactionItem>,
) {
    /** Fill the originals back into [processed] by substituting each placeholder. */
    fun restore(processed: String): String {
        var out = processed
        for (item in items) out = out.replace(item.placeholder, item.original)
        return out
    }
}

/** Options controlling detection and redaction. */
data class Options(
    /** Minimum confidence for neural detections. Structured recognizers always
     * apply. Default `0.6`. */
    val minimumConfidence: Double = 0.6,
    /** If set, only these categories are detected/redacted; otherwise all are. */
    val labels: Set<String>? = null,
)

/** Thrown when the model cannot be created, loaded, or run. */
class RedactException(message: String) : Exception(message)

/**
 * On-device, multilingual PII redaction. Mirrors the iOS/Swift SDK: create one
 * `Redact` and reuse it; the model loads lazily on the first [redaction] (or
 * eagerly via [download]).
 *
 * ```kotlin
 * val redact = Redact(context)                 // bundled model by default
 * val r = redact.redaction("Email Anna at anna@example.com.")
 * r.redactedText            // "Email [GIVEN_NAME_1] at [EMAIL_1]."
 * redact.close()
 * ```
 */
class Redact private constructor(private val handle: Long) : AutoCloseable {
    /**
     * A redactor using the bundled model by default. When [directory] is
     * supplied, that directory is treated as the model's home instead (adopt
     * files there, else download into it). Construction is cheap; the model
     * loads on the first [redaction] (or eagerly via [download]).
     */
    constructor(context: android.content.Context, directory: String? = null)
        : this(if (directory == null) bundledHandleOrNull() ?: createHandle(context.cacheDir.absolutePath, null)
               else createHandle(context.cacheDir.absolutePath, directory))

    companion object {
        /**
         * A redactor using the bundled model (no network). The main Redact AAR
         * depends on the resources artifact by default; this remains useful for
         * explicit offline construction.
         */
        fun bundled(): Redact {
            val handle = bundledHandleOrNull() ?: throw RedactException(
                "bundled model unavailable; make sure `ai.desertant:redact-tflite-resources` is present")
            return Redact(handle)
        }

        private fun bundledHandleOrNull(): Long? {
            RedactNative.ensureLoaded()
            val tokenizer = resourceOrNull("redact_tokenizer.bin") ?: return null
            val labels = resourceOrNull("labels.json") ?: return null
            val model = resourceOrNull("redact.tflite") ?: return null
            val handle = RedactNative.createBundled(tokenizer, labels, model)
            return handle.takeIf { it != 0L }
        }

        private fun createHandle(cacheRoot: String, directory: String?): Long {
            RedactNative.ensureLoaded()
            val handle = RedactNative.create(
                cacheRoot.toByteArray(Charsets.UTF_8), directory?.toByteArray(Charsets.UTF_8))
            if (handle == 0L) throw RedactException("failed to create Redact")
            return handle
        }

        private fun resourceOrNull(name: String): ByteArray? =
            Redact::class.java.getResourceAsStream("/$name")?.use { it.readBytes() }
    }

    /** Whether the model is available for this redactor with no network. */
    fun isDownloaded(): Boolean = RedactNative.isDownloaded(handle) != 0

    /**
     * Download the model ahead of time so the first [redaction] is instant. A
     * no-op once available (see [isDownloaded]). Suspends on a background
     * dispatcher.
     */
    suspend fun download(): Unit = withContext(Dispatchers.IO) {
        if (RedactNative.download(handle) != 0) throw RedactException("model download failed")
    }

    /**
     * Detect and redact the PII in [text]. Each entity is replaced by a unique,
     * numbered placeholder (`[EMAIL_1]`, `[GIVEN_NAME_1]`, ...), safe to hand to
     * an LLM and restore afterwards via [Redaction.restore]. Loads the model
     * lazily on first call.
     */
    suspend fun redaction(text: String, options: Options = Options()): Redaction =
        withContext(Dispatchers.Default) {
            val bytes = RedactNative.run(
                handle, text.toByteArray(Charsets.UTF_8), options.minimumConfidence, csv(options.labels))
                ?: throw RedactException("redaction failed")
            val buf = FfiReader(bytes)  // matches the native FFIWriter encoding
            val redactedText = buf.string()
            val n = buf.int()
            val items = ArrayList<RedactionItem>(n)
            repeat(n) {
                items.add(
                    RedactionItem(
                        label = buf.string(),
                        original = buf.string(),
                        placeholder = buf.string(),
                        confidence = buf.double(),
                        start = buf.int(),
                        end = buf.int(),
                    )
                )
            }
            Redaction(redactedText = redactedText, items = items)
        }

    /** Release the native model. The redactor is unusable afterwards. */
    override fun close() = RedactNative.destroy(handle)

    private fun csv(labels: Set<String>?): ByteArray? =
        labels?.joinToString(",")?.toByteArray(Charsets.UTF_8)
}
